tf/bonsai/bonsai.py

```diff
@@ -0,0 +1,210 @@
+import utils
+import tensorflow as tf
+import numpy as np
+
+## Bonsai Class
+class Bonsai:
+	## Constructor
+	def __init__(self, C, F, P, D, S, lW, lT, lV, lZ, 
+		sW, sT, sV, sZ, lr = None, W = None, T = None, 
+		V = None, Z = None, feats = None):
+
+		self.dataDimension = F + 1
+		self.projectionDimension = P
+		
+		if (C > 2):
+			self.numClasses = C
+		elif (C == 2):
+			self.numClasses = 1
+
+		self.treeDepth = D
+		self.sigma = S
+		
+		## Regularizer coefficients
+		self.lW = lW
+		self.lV = lV
+		self.lT = lT
+		self.lZ = lZ
+		
+		## Sparsity hyperparams
+		self.sW = sW
+		self.sV = sV
+		self.sT = sT
+		self.sZ = sZ
+
+		self.internalNodes = 2**self.treeDepth - 1
+		self.totalNodes = 2*self.internalNodes + 1
+
+		## The Parameters of Bonsai
+		self.W = self.initW(W)
+		self.V = self.initV(V)
+		self.T = self.initT(T)
+		self.Z = self.initZ(Z)
+
+		## Placeholders for Hard Thresholding and Sparse Training
+		self.Wth = tf.placeholder(tf.float32, name='Wth')
```

Comment:
@adityakusupati  We should move anything related to  training to a separate method and not make it part of the `__init__`. Sometimes, you'd just want to have a Bonsai subgraph initialized with pretrained weights for inference. For such cases, there is no reason to have these unused nodes (`Wth, Vth` etc) as part of the graph.

Also, `__init__` should also take the graph to attach to as an argument. What if I don't want Bonsai to be part of the global graph and instead am running multiple experimetns with different graphs ?

Reply:
I agree with the first part. I need to think about it.

The second part is already addressed. feats is the argument which takes in the graph to be attached. I didn't get the doubt.